mne/time_frequency/csd.py

```diff
@@ -255,3 +220,201 @@ def compute_epochs_csd(epochs, mode='multitaper', fmin=0, fmax=np.inf,
+        in the time domain. If True, a single CSD matrix will be returned. If
+        False, the output will be a numpy.ndarray of CSD matrices.
+    n_fft : int | None
+        Length of the FFT. If None the exact number of samples between tmin and
+        tmax will be used.
+    mt_bandwidth : float | None
+        The bandwidth of the multitaper windowing function in Hz.
+        Only used in 'multitaper' mode.
+    mt_adaptive : bool
+        Use adaptive weights to combine the tapered spectra into PSD.
+        Only used in 'multitaper' mode.
+    mt_low_bias : bool
+        Only use tapers with more than 90% spectral concentration within
+        bandwidth. Only used in 'multitaper' mode.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    csd : numpy.ndarray
+        The computed cross spectral-density (either summed or not).
+    freqs : numpy.ndarray
+        Frequencies the cross spectral-density is evaluated at.
+    """
+
+    # Check correctness of input data and parameters
+    if fmax < fmin:
+        raise ValueError('fmax must be larger than fmin')
+
+    if isinstance(X, np.ndarray):
+        if X.ndim == 3:
+            n_trials, n_series, n_times = X.shape
+        elif X.ndim == 2:
+            n_trials = 1
+            n_series, n_times = X.shape
+        else:
+            err_str = "X must be either n_trials x n_seris x n_times "
+            err_str += "or n_series x n_times ndarray"
+            raise ValueError(err_str)
+    elif isinstance(X, list):
+        if any([not isinstance(x, np.ndarray) for x in X]):
+            raise ValueError("All arrays in X must be ndarrays")
+        shapes = set(x.shape for x in X)
+        if len(shapes) != 1:
+            raise ValueError("All arrays in X must be the same shape")
+        n_trials = len(X)
+        n_series, n_times = X[0].shape
+    n_fft = n_times if n_fft is None else n_fft
+
+    # Preparing frequencies of interest
+    orig_frequencies = fftfreq(n_fft, 1. / sfreq)
+    freq_mask = (orig_frequencies > fmin) & (orig_frequencies < fmax)
+    frequencies = orig_frequencies[freq_mask]
+    n_freqs = len(frequencies)
+
+    if n_freqs == 0:
+        raise ValueError('No discrete fourier transform results within '
+                         'the given frequency window. Please widen either '
+                         'the frequency window or the time window')
+
+    # Preparing for computing CSD
+    logger.info('Computing cross-spectral density from ndarray...')
+    pars = _compute_csd_params(n_times, sfreq, mode, mt_bandwidth, mt_low_bias,
+                               mt_adaptive)
+    window_fun, eigvals, n_tapers, mt_adaptive = pars
+
+    csds_mean = np.zeros((n_series, n_series, n_freqs), dtype=complex)
+
+    # Picking frequencies of interest
+    freq_mask_mt = freq_mask[orig_frequencies >= 0]
+
+    # Compute CSD for each trial
+    for ti in xrange(n_trials):
```

Comment:
`xrange` does not exist on Py3k. `n_trails` will probably never be huge so just use `range` here. In the future if you really need it you can get it from `six` (which we keep a copy of in `externals`)